src/org/traccar/protocol/Tk103ProtocolDecoder.java

```diff
@@ -111,38 +132,140 @@ private void decodeType(Position position, String type, String data) {
+            case "DW42":
+            case "DW62":
+                position.set(Position.KEY_IGNITION, true);
+                break;
+            // Dismounting Alert with Location
             case "ZC17":
+            case "DW37":
+            case "DW57":
                 position.set(Position.KEY_ALARM, Position.ALARM_REMOVING);
                 break;
+            // SOS Alert with Location
             case "ZC25":
+            case "DW3E":
+            case "DW5E":
                 position.set(Position.KEY_ALARM, Position.ALARM_SOS);
                 break;
+            // Tampering Alert with Location:
             case "ZC26":
+            case "DW3F":
+            case "DW5F":
                 position.set(Position.KEY_ALARM, Position.ALARM_TAMPERING);
                 break;
+            // Low Power Alert with Location
             case "ZC27":
+            case "DW40":
+            case "DW60":
                 position.set(Position.KEY_ALARM, Position.ALARM_LOW_POWER);
                 break;
+            // Bad Battery Alert with Location
+            case "ZC28":
+            case "DW41":
+            case "DW61":
+                position.set(Position.KEY_ALARM, "badBattery");
+                break;
+            // Fuel Cut On with Location
+            case "ZC2A":
+            case "DW43":
+            case "DW63":
+                position.set(Position.KEY_FUEL_LEVEL, 0.0);
+                break;
+            // Fuel Cut Off with Location:
+            case "ZC2B":
+            case "DW44":
+            case "DW64":
+                position.set(Position.KEY_FUEL_LEVEL, 1.0);
+                break;
+            // Fuel Cut Alert with Location:
+            case "ZC2C":
+            case "DW45":
+            case "DW65":
+                position.set(Position.KEY_ALARM, "fuelCut");
+                break;
             default:
                 break;
         }
     }
 
+    private Integer decodeBattery(int value) {
+        switch (value) {
+            case 6:
+                return 100;
+            case 5:
+                return 80;
+            case 4:
+                return 60;
+            case 3:
+                return 20;
+            case 2:
+                return 10;
+            default:
+                return null;
+        }
+    }
+
+    private static Map<Long, Position> batteryInfo = new HashMap<Long, Position>();
+
+    public void setLastBatteryPower(Position positionWithCurrentBatteryInfo) {
+        batteryInfo.put(positionWithCurrentBatteryInfo.getDeviceId(), positionWithCurrentBatteryInfo);
+    }
+
+    // It is possible to use "processing.copyAttributes" configuration to have battery level in all events,
+    //  but this will not work if battery events will be filtered somehow and dropped from DB, so it will be no
+    //  event in DB with latest battery level. So, ensure all our events have latest battery level.
+    public void getLastBatteryPower(Position position) {
```

Comment:
This has to be removed. Copy attributes functionality covers it.

Reply:
even if battery events will be filtered? (they have timestamp (fixTime) same as previous event (because of getLastLocation()), so alway filtered by "filter.duplicate")

Reply:
There is an option to not filter messages with certain attributes. Either way it's not a good idea to implement some workaround for a single protocol.

Reply:
ok, i will remove...